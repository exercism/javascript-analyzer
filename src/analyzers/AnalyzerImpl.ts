import { getProcessLogger as getLogger, Logger } from '../utils/logger'

import { AnalyzerOutput } from '../output/AnalyzerOutput';
import { ParsedSource, AstParser } from '../parsers/AstParser';

class EarlyFinalization extends Error {
  constructor() {
    super('Early finalization')
    Object.setPrototypeOf(this, EarlyFinalization.prototype);
    Error.captureStackTrace(this, this.constructor)
  }
}

export abstract class AnalyzerImpl implements Analyzer {
  protected readonly logger: Logger
  private output!: AnalyzerOutput

  /**
   * Creates an instance of an analyzer
   */
  constructor() {
    this.logger = getLogger()
  }

  /**
   * Runs the analyzer
   *
   * This is defined as a property instead of a method, so that it can not be
   * overriddden in a subclass. Subclasses should override @see execute instead.
   *
   * @returns The promise that resolves the analyzer output.
   *
   * @memberof BaseAnalyzer
   */
  public async run(input: Input): Promise<Output> {
    // Ensure each run has a fresh output
    //
    // Note: still need to wait for a run to complete before the next one can be
    //       started. We could work around this by providing an execution
    //       context that is fresh on each run.
    //
    // The reason output is not passed to execute, is that it doesn't _actually_
    // enforce the implementing analyzer to not use local state, so we don't
    // gain anything by it.
    //
    this.output = new AnalyzerOutput()

    await this.execute(input)
      .catch((err) => {
        if (err instanceof EarlyFinalization) {
          this.logger.log(`=> early finialization (${this.output.status})`)
        } else {
          throw err
        }
      })

    return this.output
  }

  /**
   * Approve the solution early with an optional comment.
   *
   * @see disapprove
   * @see redirect
   *
   * @param comment the optional comment to approve with
   * @throws {EarlyFinalization} used as control flow in @see run
   */
  protected approve(comment?: Comment): never {
    this.comment(comment)
    this.output.approve()

    throw new EarlyFinalization()
  }

  /**
   * Disapprove the solution early with an optional comment
   *
   * @see approve
   * @see redirect
   *
   * @param comment the optional comment to disapprove with
   * @throws {EarlyFinalization} used as control flow in @see run
   */
  protected disapprove(comment?: Comment) {
    this.comment(comment)
    this.output.disapprove()

    throw new EarlyFinalization()
  }

  /**
   * Refer the solution to the mentor early with an optional comment
   *
   * @see approve
   * @see disapprove
   *
   * @param comment the optional comment to redirect with
   * @throws {EarlyFinalization} used as control flow in @see run
   */
  protected redirect(comment?: Comment): never {
    this.comment(comment)
    this.output.redirect()

    throw new EarlyFinalization()
  }

  /**
   * Add a comment to the output
   *
   * @param {Comment} [comment]
   */
  protected comment(comment?: Comment) {
    if (!comment) {
      return
    }

    this.output.add(comment)
  }

  /**
   * Property that returns true if there is at least one comment in the output.
   */
  get hasCommentary() {
    return this.output.comments.length > 0
  }

  /**
   * Execute the analyzer
   */
  protected abstract execute(input: Input): Promise<void>
}
