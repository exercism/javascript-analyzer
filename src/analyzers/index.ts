import { BaseAnalyzer } from './base_analyzer'

import { Exercise } from '../exercise'
import { Solution } from '../solution'

import { get as getLogger } from '../utils/logger'

import path from 'path'

export class Analyzers {

  /**
   * Find the correct analyzer given an exercise
   *
   * @param exercise the exercise to lookup
   *
   * @throws if there is no such analyzer
   * @returns the constructor for the analyzer
   */
  public static find<T extends BaseAnalyzer>(exercise: Readonly<Exercise>): new (solution: Solution) => T  {
    const file = Analyzers.autoload(exercise)
    const key = Object.keys(file).find(key => file[key] instanceof Function)

    if (key === undefined) {
      throw new Error(`No Analyzer found in './${exercise.slug}`)
    }

    const analyzer = file[key]
    getLogger().log(`=> analyzer: ${analyzer.name}`)
    return analyzer
  }

  private static autoload(exercise: Readonly<Exercise>) {
    const modulePath = path.join(__dirname, exercise.slug, 'index') // explicit path (no extension)
    try {
      return require(modulePath)
    } catch(err) {
      const logger = getLogger()
      logger.error(`
Could not find the index.js analyzer in "${modulePath}"

Make sure that:
 - the slug "${exercise.slug}" is valid (hint: use dashes, not underscores)
 - there is actually an analyzer written for that exercise

Original error:

`.trimLeft())
      logger.fatal(JSON.stringify(err), -32)
    }
  }
}
