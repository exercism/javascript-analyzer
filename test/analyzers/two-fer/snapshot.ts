import { TwoFerAnalyzer } from '~src/analyzers/two-fer'
import { makeTestGenerator } from '~test/helpers/snapshot'

const snapshotTestsGenerator = makeTestGenerator(
  'two-fer',
  () => new TwoFerAnalyzer()
)

describe('When running analysis on two-fer fixtures', () => {
  snapshotTestsGenerator('approve', [
    129, 139, 400, 493, 70
  ])
  snapshotTestsGenerator('disapprove', [
    1, 10, 100, 101, 102, 104, 105, 107, 109, 11,
    110, 111, 112, 114, 115, 116, 12, 121, 123, 124,
    118, 120, 313
  ])
  snapshotTestsGenerator('refer_to_mentor', [
    0, 103, 106, 108, 113, 119, 122, 133, 138,
    143, 147, 148, 166, 171, 175, 181, 183, 192, 194
  ])
})
