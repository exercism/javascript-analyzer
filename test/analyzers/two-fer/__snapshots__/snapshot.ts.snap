// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/70's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/118's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/129's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/139's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/313's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/400's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve matches two-fer/493's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/1's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/10's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/11's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.generic.no_method",
      "message": "No method called \`twoFer\`. The tests won't pass without it.",
      "template": "No method called \`%<method.name>s\`. The tests won't pass without it.",
      "variables": Object {
        "method.name": "twoFer",
      },
    },
    CommentImpl {
      "externalTemplate": "javascript.generic.no_named_export",
      "message": "No [export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) called \`twoFer\`.
The tests won't pass without it.

Did you forget adding: \`export twoFer\`?",
      "template": "No [export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) called \`%<export.name>s\`.
The tests won't pass without it.

Did you forget adding: \`export %<export.name>s\`?",
      "variables": Object {
        "export.name": "twoFer",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/12's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/100's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/101's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.generic.prefer_strict_equality",
      "message": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "template": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "variables": Object {},
    },
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/102's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.generic.prefer_strict_equality",
      "message": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "template": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "variables": Object {},
    },
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/104's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of X to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "X",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/105's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_explicity_default_value",
      "message": "Instead of relying on %<maybe_undefined_expression>s being \\"undefined\\" when
no value is passed in, you could set the default value of 'name' to
'you'.",
      "template": "Instead of relying on %<maybe_undefined_expression>s being \\"undefined\\" when
no value is passed in, you could set the default value of '%<parameter>s' to
'you'.",
      "variables": Object {
        "maybe-undefined-expression": "name",
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/107's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_explicity_default_value",
      "message": "Instead of relying on %<maybe_undefined_expression>s being \\"undefined\\" when
no value is passed in, you could set the default value of 'name' to
'you'.",
      "template": "Instead of relying on %<maybe_undefined_expression>s being \\"undefined\\" when
no value is passed in, you could set the default value of '%<parameter>s' to
'you'.",
      "variables": Object {
        "maybe-undefined-expression": "name",
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/109's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/110's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.generic.prefer_strict_equality",
      "message": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "template": "In _JavaScript_, always prefer [strict (identity and non-identity) equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity)
such as \`===\` and \`!==\` over the forms that use implicit type coercion,
such as [\`==\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Equality)
and [\`!=\`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Inequality),
unless you explicitly want to coerce the type of one of the two operands.

There are definitely cases where you'll want to use non-strict equality, but
that's not the case in this exercise.",
      "variables": Object {},
    },
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/111's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/112's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.generic.no_method",
      "message": "No method called \`twoFer\`. The tests won't pass without it.",
      "template": "No method called \`%<method.name>s\`. The tests won't pass without it.",
      "variables": Object {
        "method.name": "twoFer",
      },
    },
    CommentImpl {
      "externalTemplate": "javascript.generic.no_named_export",
      "message": "No [export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) called \`twoFer\`.
The tests won't pass without it.

Did you forget adding: \`export twoFer\`?",
      "template": "No [export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) called \`%<export.name>s\`.
The tests won't pass without it.

Did you forget adding: \`export %<export.name>s\`?",
      "variables": Object {
        "export.name": "twoFer",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/114's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/115's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/116's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/121's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/123's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to disapprove matches two-fer/124's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.optimise_default_value",
      "message": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of name to 'you' to avoid this conditional.",
      "template": "You currently use a conditional to branch in case there is no value passed into
\`twoFer\`. Instead you could set the [default value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters)
of %<parameter>s to 'you' to avoid this conditional.",
      "variables": Object {
        "parameter": "name",
      },
    },
  ],
  "status": "disapprove",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/0's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/103's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/106's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/108's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/113's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/119's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/120's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [
    CommentImpl {
      "externalTemplate": "javascript.two-fer.redirect_incorrect_string_template",
      "message": "The string template looks incorrect. Expected a template with 3 components.",
      "template": "The string template looks incorrect. Expected a template with 3 components.",
      "variables": Object {},
    },
  ],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/122's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/133's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/138's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/143's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/147's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/148's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/166's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/171's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/175's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/181's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/183's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/192's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to refer to mentor matches two-fer/194's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "refer_to_mentor",
}
`;
